python/ray/data/_internal/planner/plan_udf_map_op.py

```diff
@@ -280,17 +305,182 @@ def plan_udf_map_op(
+    partition_transform_fn = _generate_transform_fn_for_map_batches(fn)
+    partition_transform_fns = [
+        # Convert input blocks to batches.
+        BlocksToBatchesMapTransformFn(
+            batch_size=None,  # TODO: Should we use op._batch_size here?
+            batch_format="pyarrow",
+            zero_copy_batch=True,
+        ),
+        # Apply the PartitionActor function.
+        BatchMapTransformFn(partition_transform_fn),
+    ]
+    partition_map_transformer = MapTransformer(partition_transform_fns, init_fn)
+    partition_map_operator = MapOperator.create(
+        partition_map_transformer,
+        input_physical_dag,
+        data_context,
+        name="DownloadURIPartitioner",
+        compute_strategy=compute,
+        ray_remote_args=op._ray_remote_args,
+        ray_remote_args_fn=op._ray_remote_args_fn,
+    )
+
+    fn, init_fn = _get_udf(download_images_threaded, (uri_column_name,), {}, (), {})
+    download_transform_fn = _generate_transform_fn_for_map_batches(fn)
+    transform_fns = [
+        # Apply the download function.
+        BatchMapTransformFn(download_transform_fn),
+        # Convert output batches to blocks.
+        BuildOutputBlocksMapTransformFn.for_batches(),
+    ]
+    download_map_transformer = MapTransformer(transform_fns, init_fn)
+    download_map_operator = MapOperator.create(
+        download_map_transformer,
+        partition_map_operator,
+        data_context,
+        name="DownloadURIDownloader",
+        compute_strategy=compute,
+        ray_remote_args=op._ray_remote_args,
+        ray_remote_args_fn=op._ray_remote_args_fn,
+    )
+
+    return download_map_operator
+
+
+def download_images_threaded(batch, uri_column_name, max_workers=16):
+    """Optimized version that uses make_async_gen for concurrent downloads."""
+    from pyarrow import fs as pafs
+
+    from ray.data._internal.util import make_async_gen
+
+    def download_uris(uri_iterator):
+        """Function that takes an iterator of URIs and yields downloaded bytes for each."""
+        # TODO: We need to update this to use pafs.from_uri here.
+        fs = pafs.S3FileSystem(region="us-west-2")
+        for uri in uri_iterator:
+            try:
+                with fs.open_input_file(uri) as f:
+                    yield f.read()
+            except Exception as e:
+                print(f"Error downloading {uri}: {e}")
+                yield None
+
+    # Extract URIs from PyArrow table
+    uris = batch.column(uri_column_name).to_pylist()
+
+    # Use make_async_gen to download URIs concurrently
+    # This preserves the order of results to match the input URIs
+    images = list(
+        make_async_gen(
+            base_iterator=iter(uris),
+            fn=download_uris,
+            preserve_ordering=True,
+            num_workers=max_workers,
+        )
+    )
+
+    # Add the new column to the PyArrow table
+    return batch.add_column(len(batch.column_names), "image_bytes", pa.array(images))
```

Comment:
I think this is supposed to be the user-specific column name in `with_columns`?
```suggestion
    return batch.add_column(len(batch.column_names), "image_bytes", pa.array(images))
```